src/runtime_src/core/common/pskernel_parse.cpp

```diff
@@ -0,0 +1,230 @@
+	  // Extract parameter name and type
+	  xrt_core::pskernel::kernel_argument arg;
+	  var_name = dwarf_diename(&child);
+	  arg.name = var_name;
+	  type = dwarf_formref_die(dwarf_attr(&child, DW_AT_type, &attr_mem),&type_mem);
+	  while(dwarf_tag(type) == DW_TAG_typedef) {
+	    type = dwarf_formref_die(dwarf_attr(type, DW_AT_type, &attr_mem),&type_mem);
+	  }
+	  if(dwarf_tag(type) == DW_TAG_base_type) {
+	    Dwarf_Attribute encoding;
+	    Dwarf_Word enctype = 0;
+	    if (dwarf_attr (type, DW_AT_encoding, &encoding) == NULL
+		|| dwarf_formudata (&encoding, &enctype) != 0)
+	      throw std::runtime_error("base type without encoding");
+	    
+	    Dwarf_Attribute bsize;
+	    Dwarf_Word bits;
+	    if (dwarf_attr (type, DW_AT_byte_size, &bsize) != NULL
+		&& dwarf_formudata (&bsize, &bits) == 0) {
+	      arg.size = bits;
+	    } else if (dwarf_attr (type, DW_AT_bit_size, &bsize) == NULL
+		     || dwarf_formudata (&bsize, &bits) != 0)
+	      throw std::runtime_error("base type without byte or bit size");
+
+	    auto result = typeTable.find(std::make_pair(enctype,bits));
+	    ffi_type *enc_ffi_type = result->second;
+	    switch(enctype) {
+	      case DW_ATE_signed_char:
+		arg.hosttype = "int8_t";
+		break;
+	      case DW_ATE_unsigned_char:
+		arg.hosttype = "uint8_t";
+		break;
+	      case DW_ATE_signed:
+		if(bits==2)
+		  arg.hosttype = "int16_t";
+		if(bits==4)
+		  arg.hosttype = "int";
+		if(bits==8)
+		  arg.hosttype = "int64_t";
+		break;
+	      case DW_ATE_unsigned:
+		if(bits==2)
+		  arg.hosttype = "uint16_t";
+		if(bits==4)
+		  arg.hosttype = "uint32_t";
+		if(bits==8)
+		  arg.hosttype = "uint64_t";
+		break;
+	      case DW_ATE_float:
+		if(bits==4)
+		  arg.hosttype = "float";
+		if(bits==8)
+		  arg.hosttype = "double";
+		break;
+	    }
+	    arg.ffitype = *enc_ffi_type;
+	    arg.offset = offset;
+	    arg.type = xrt_core::pskernel::kernel_argument::argtype::scalar;
+	    offset = offset + bits;
+	    arg.index = index;
+	    index++;
+	  } else if (dwarf_tag(type) == DW_TAG_pointer_type) {
+	    arg.size = 16;
+	    arg.offset = offset;
+	    offset = offset + 16;  // Next offset increase by 3 words to account for 64-bit address and 64-bit size
+	    arg.index = index;
+	    arg.type = xrt_core::pskernel::kernel_argument::argtype::global;
+	    ffi_type *enc_ffi_type = &ffi_type_pointer;
+	    arg.ffitype = *enc_ffi_type;
+	    index++;
+	  }
+	  if(dwarf_aggregate_size(type, &var_size) < 0)
+	    throw std::runtime_error("ERROR: Variable Size incorrect!\n");
+
+	  return_args.emplace_back(arg);
+	  break;
+	} while (dwarf_siblingof (&child, &child) == 0);
+
+  return return_args;
+}
+
+// Function to parse object from file on disk
+std::vector<xrt_core::pskernel::kernel_argument> pskernel_parse(const char *so_file, const char *func_name) {
+
+  int fd = open(so_file, O_RDONLY);
+  Dwarf *dw = dwarf_begin(fd, DWARF_C_READ);
+  std::vector<xrt_core::pskernel::kernel_argument> args;
+
+  if (dw != NULL) {
```

Comment:
No NULL is C++. 
Can the logic be reverted to do early-exit and avoid the indentation?